File: go/src/2020/july/all-paths-from-source-to-target.go
package july

func all_paths(graph [][]int) [][]int {

	var res = [][]int{}
	dfs(graph, 0, res, []int{0})
	return res
}

func dfs(graph [][]int, cur int, res [][]int, path []int) {
	if cur == len(graph)-1 {
		res = append(res, path)
		return
	}
	for _, next := range graph[cur] {
		path = append(path, next)
		dfs(graph, next, res, path)
		path = path[0 : len(path)-1]
	}
}

File: code/python/2020/july/all-paths-from-source-to-target.py
def allPathsSourceTarget(graph):
    def dfs(cur, path):
        if cur == len(graph) - 1:
            res.append(path)
        else:
            for i in graph[cur]: dfs(i, path + [i])

    res = []
    dfs(0, [0])
    return res


def _allPathsSourceTarget(self, g, cur=0):
    if cur == len(g) - 1: return [[len(g) - 1]]
    return [([cur] + path) for i in g[cur] for path in self._allPathsSourceTarget(g, i)]
